### ISSUE TITLE
Questions about the plots

### ISSUE THREAD

**Issue by LightingMc** [eladrich/pixel2style2pixel#155]
I hope these plots below make sense. 

![image](https://user-images.githubusercontent.com/42900802/122814964-e3019a00-d2a2-11eb-8277-4ea6ed6e8aac.png)

1. is the "loss_w_norm" supposed to increase with the epochs? 

2. Is the "id_improve" supposed to be negative and increase with time as well?

**Comment by yuval-alaluf** [comment]
Hi @LightingMc ,
To answer your questions: 
1. The `w_norm` loss can increase during training. This does make sense because during training the encoder will move its encodings further away from the latent average, therefore increasing the loss. So this should be fine. 
2. The `id_improve` should be negative and increase with time. A large negative `id_improve` value means there's a large difference in `id` between the target and output while an `id_improve` of 0 means they are equal.
In other words, all your graphs look good :)

**Comment by LightingMc** [comment]
Awesome. Thanks.

**Comment by LightingMc** [comment]
There were some issues with uploading an image, but I added an image of a new loss plot below.

[https://photos.app.goo.gl/Psnbn7wMu5EEPGSg8](url)

The training ID_improve goes above 0, it becomes positive. That is a bad thing right?

**Comment by yuval-alaluf** [comment]
Regarding (2), this looks good actually. In the beginning, you're moving away from the average latent code and so the loss increases. But during training, the w-norm loss starts to work and pushes the latents back toward the average latent code and you see the decrease there at the end. So the loss, in my opinion, is working exactly as it should :)

Regarding (1), in none of my experiments did I see `id_improve` above 0. However, the `id_improve` is pretty much computed as `y_hat_feats[i].dot(y_feats[i])` which is the cosine similarity between the GT face embeddings and predicted face embeddings. And this `dot` operation can return values which are greater than 1. So I do think that it is possible to get values greater than 0. What is more interesting to me is the large gap between the train `id_improve` and the test `id_improve`. I am not sure I ever saw a gap that big. 

In any case, to summarize, both losses seem good, but I am a bit curious as to the large gap in the `id_improve` between the train and test. 

If I may ask, are the results you're getting looking good? Or do you suspect there is something wrong with the training resulting in poor results?

**Comment by LightingMc** [comment]
Hey, sorry for the late reply.  Some issues in my lab's server because of which I couldn't access the results for a while. 

Actually, the results that we are getting are good enough but not that high quality. My advisors said that the results should be of much better quality since we are using StyleGAN. Do you have any suggestions as to why?

![image](https://user-images.githubusercontent.com/42900802/124200339-1a422900-daa3-11eb-9def-223ffae5d720.png)

This thing was trained for LR 0.0001, w_norm_lambda 0.0005, UnNormalised input and out, and started from the average latent code. The loss plot is below:
![image](https://user-images.githubusercontent.com/42900802/124200613-cab02d00-daa3-11eb-95f0-47977cfe58af.png)


Regards,
Zaigham

**Comment by yuval-alaluf** [comment]
Are you trying to train the inversion task? If so, it seems like the reconstructions are working great, but it seems like the target images are simply not loaded correctly (they're dulled out). 
Are you sure you're loading the images correctly?

**Comment by LightingMc** [comment]
1. Yes, I am trying to train the inversion task (however the StyleGAN is frozen by default in your code I think, haven't touched that yet). 

2. The blur is because of normalization of the inputs.

**Comment by LightingMc** [comment]
1. Yes, I am trying to train the inversion task (however the StyleGAN is frozen by default in your code I think, haven't touched that yet).

2. The dull is because of the normalization of the inputs.

![image](https://user-images.githubusercontent.com/42900802/124377043-ec620d80-dc77-11eb-837f-a883f47d860e.png)


Some results with the non-normalised (non-dulled images) can be seen below:

![image](https://user-images.githubusercontent.com/42900802/124377143-71e5bd80-dc78-11eb-8a99-911ac5c52929.png)

**Comment by LightingMc** [comment]
3. Do u think any improvement on these results can be obtained?
4. There is something up with the eyes. For all of the results, they look similar. Look at the lower person in the image above, the eyes are looking one way in the target, but the output image's eyes are like in the dead center.

**Comment by yuval-alaluf** [comment]
> 1. Do u think any improvement on these results can be obtained?

The inversion look quite good and the identity preservation looks excellent. If you feel like this is not enough, you can try out a more recent work of mine called ReStyle which improves on the inversion of pSp by a decent margin. Here's the link to the repo: https://github.com/yuval-alaluf/restyle-encoder

> 2. There is something up with the eyes. For all of the results, they look similar. Look at the lower person in the image above, the eyes are looking one way in the target, but the output image's eyes are like in the dead center.

I would say that this would fall under the category of "fine details" which as you mention is quite hard to preserve. A more accurate inversion method may be able to help here so ReStyle may be able to assist with this as well

**Comment by LightingMc** [comment]
Awesome, thanks I will check it out.
